pkg/channel/http/http_channel.go

```diff
@@ -65,116 +62,103 @@ func (h *Channel) GetBaseAddress() string {
 }
 
 // InvokeMethod invokes user code via HTTP
-func (h *Channel) InvokeMethod(invokeRequest *channel.InvokeRequest) (*channel.InvokeResponse, error) {
-	req := fasthttp.AcquireRequest()
-	uri := fmt.Sprintf("%s/%s", h.baseAddress, invokeRequest.Method)
-	req.SetRequestURI(uri)
-	req.SetBody(invokeRequest.Payload)
-	if val, ok := invokeRequest.Metadata[QueryString]; ok {
-		req.URI().SetQueryString(val)
+func (h *Channel) InvokeMethod(req *invokev1.InvokeMethodRequest) (*invokev1.InvokeMethodResponse, error) {
+	// Check if HTTP Extension is given. Otherwise, it will return error.
+	httpExt := req.Message().GetHttpExtension()
+	if httpExt == nil {
+		return nil, status.Error(codes.InvalidArgument, "missing HTTP extension field")
 	}
-	if invokeRequest.Metadata != nil {
-		if val, ok := invokeRequest.Metadata["headers"]; ok {
-			headers := strings.Split(val, HeaderDelim)
-			for _, h := range headers {
-				kv := strings.Split(h, HeaderEquals)
-				req.Header.Set(kv[0], kv[1])
-			}
-		}
+	if httpExt.GetVerb() == commonv1pb.HTTPExtension_NONE {
+		return nil, status.Error(codes.InvalidArgument, "invalid HTTP verb")
 	}
-	applyContentTypeIfNotPresent(req)
 
-	method := invokeRequest.Metadata[HTTPVerb]
-	if method == "" {
-		method = Post
-	}
-	req.Header.SetMethod(method)
+	var rsp *invokev1.InvokeMethodResponse
+	var err error
+	switch req.APIVersion() {
+	case internalv1pb.APIVersion_V1:
+		rsp, err = h.invokeMethodV1(req)
 
-	if invokeRequest.Metadata != nil {
-		if corID, ok := invokeRequest.Metadata[diag.CorrelationID]; ok {
-			req.Header.Set(diag.CorrelationID, corID)
-		}
+	default:
+		// Reject unsupported version
+		err = status.Error(codes.Unimplemented, fmt.Sprintf("Unsupported spec version: %d", req.APIVersion()))
 	}
 
-	var span diag.TracerSpan
-	var spanc diag.TracerSpan
-
-	span, spanc = diag.TraceSpanFromFastHTTPRequest(req, h.tracingSpec)
-	defer span.Span.End()
-	defer spanc.Span.End()
-
-	req.Header.Set(diag.CorrelationID, diag.SerializeSpanContext(*spanc.SpanContext))
+	return rsp, err
+}
 
-	resp := fasthttp.AcquireResponse()
+func (h *Channel) invokeMethodV1(req *invokev1.InvokeMethodRequest) (*invokev1.InvokeMethodResponse, error) {
+	channelReq := h.constructRequest(req)
 
 	if h.ch != nil {
 		h.ch <- 1
 	}
 
+	// Emit metric when request is sent
 	// TODO: Use propagated context
 	ctx := context.Background()
-	// Emit metric when request is sent
-	diag.DefaultHTTPMonitoring.ClientRequestStarted(
-		ctx, method, invokeRequest.Method,
-		int64(len(invokeRequest.Payload)))
-
+	diag.DefaultHTTPMonitoring.ClientRequestStarted(ctx, req.Message().Method, req.Message().Method, int64(len(req.Message().Data.GetValue())))
```

Comment:
Can we skip every diag message if the relevant telemetry isn't enabled?

Reply:
[ClientRequestStarted](https://github.com/dapr/dapr/blob/7b5eaf3d268b1e659a0fde18ae2fb11c2b569978/pkg/diagnostics/http_monitoring.go#L111) checks if metric is enabled or not. if it is disabled, it will not emit any metric.